### PR TITLE
[package-upgrades] Added cross-version conflicting dependency call test

### PR DESCRIPTION
**Pull request by awelc**
## Description 

Another package upgrade test plus a refinement to the previous one promised on the earlier PR (https://github.com/MystenLabs/sui/pull/10788#discussion_r1164604573)

## Test Plan 

These are the actual tests

### CODE REVIEW

File: crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs [hunk 1]
```diff
@@ -1085,6 +1086,21 @@ async fn test_upgraded_types_in_one_txn() {
...
+    let e1_type = StructTag::from_str(&format!("{package_v2}::base::BModEvent")).unwrap();
+    let e2_type = StructTag::from_str(&format!("{package_v3}::base::CModEvent")).unwrap();
+
+    let event_digest = effects.events_digest.unwrap();
+    let mut events = runner
+        .authority_state
+        .get_transaction_events(&event_digest)
+        .unwrap()
+        .data;
+    events.sort_by(|a, b| a.type_.name.as_str().cmp(b.type_.name.as_str()));
```
Comment (amnn) [inline crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs:1100]: nit: `sort_by_key`
```suggestion
    events.sort_by_key(|a| a.type_.name);
```

File: crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs [hunk 2]
```diff
@@ -1114,3 +1130,108 @@ async fn test_different_versions_across_calls() {
 
     assert!(effects.status.is_ok(), "{:#?}", effects.status);
 }
+
+#[tokio::test]
+async fn test_conflicting_versions_across_calls() {
+    // publishes base package at version 1
+    let mut runner = UpgradeStateRunner::new("move_upgrade/base").await;
+
+    // publish a depender package at version 1 that depends on the base package at version 1
```
Comment (amnn) [inline crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs:1139]: nit: `depender` -> `dependent` through-out.

```suggestion
    // publish a dependent package at version 1 that depends on the base package at version 1
```

File: crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs [hunk 3]
```diff
@@ -1114,3 +1130,108 @@ async fn test_different_versions_across_calls() {
...
+        .find(|(_, owner)| matches!(owner, Owner::Immutable))
+        .unwrap()
+        .0;
+
+    // publish a depender package at version 2 that depends on the base package at version 2
+    let pt2 = {
+        let mut builder = ProgrammableTransactionBuilder::new();
+        let current_package_id = depender_package.0;
+        // Now recompile the depending package with the upgraded dependency
+        // Currently doesn't work -- need to wait for linkage table to be added to the loader.
```
Comment (amnn) [inline crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs:1172]: nit: comment out of date (I'm guessing this is copypasta from another comment that's also out-of-date?)

**Review by amnn — APPROVED** [review]
Some nits, but otherwise this test looks good!

File: crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs (hunk 1 above)
Comment (awelc) [reply crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs:1099]: I tried that initially but I am hitting some borrow checker issues when using `sort_by_key` ...

File: crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs (hunk 2 above)
Comment (awelc) [reply crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs]: To my defense, the idea was to refer to the variable name, but then indeed it should be "publish depender package". In any case, changed to "a dependent package" as it's indeed the best choice!

File: crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs (hunk 3 above)
Comment (awelc) [reply crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs]: Copypasta indeed and the other test does work now of course so removed it in both cases.

File: crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs (hunk 1 above)
Comment (amnn) [reply crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs:1099]: Try borrowing the type name, the key doesn't need to be taken by value.

**Pull request merged**